node/src/tasks/msp_move_bucket.rs

```diff
@@ -0,0 +1,306 @@
+
+        let mut indexer_connection = indexer_db_pool.get().await?;
+
+        let forest_storage = self
+            .storage_hub_handler
+            .forest_storage_handler
+            .get_or_create(&event.bucket_id.as_ref().to_vec())
+            .await;
+
+        for file in shc_indexer_db::models::File::get_by_onchain_bucket_id(
+            &mut indexer_connection,
+            bucket.clone(),
+        )
+        .await?
+        {
+            info!(
+                target: LOG_TARGET,
+                "MSP: downloading file {:?} of bucket {:?}",
+                file.file_key,
+                event.bucket_id,
+            );
+
+            let file_metadata = file.to_file_metadata(bucket.clone());
+            let file_key = file_metadata.file_key::<HashT<StorageProofsMerkleTrieLayout>>();
+
+            self.storage_hub_handler
+                .file_storage
+                .write()
+                .await
+                .insert_file(file_key.clone(), file_metadata.clone())
+                .unwrap();
+
+            let chunks_count = file_metadata.chunks_count();
+
+            let mut bsp_peer_ids = file.get_bsp_peer_ids(&mut indexer_connection).await?;
+
+            // Shuffle in order to avoid consecutive requests to the same BSP node.
+            bsp_peer_ids.shuffle(&mut rand::thread_rng());
+
+            if bsp_peer_ids.is_empty() {
+                error!(
+                    target: LOG_TARGET,
+                    "No BSP peer IDs found for file {:?} of bucket {:?}",
+                    file_key, event.bucket_id,
+                );
+                continue;
+            }
+
+            // We will cycle through all the BSP peer IDs for each chunk until we successfully
+            // download the file.
+            let mut bsp_peer_ids_iter = bsp_peer_ids.iter().cycle();
+
+            for chunk in 0..chunks_count {
+                for _ in 0..DOWNLOAD_REQUEST_RETRY_COUNT {
+                    let peer_id = bsp_peer_ids_iter.next().unwrap();
+
+                    let download_request = self
+                        .storage_hub_handler
+                        .file_transfer
+                        .download_request(
+                            *peer_id,
+                            file_key.into(),
+                            ChunkId::new(chunk),
+                            Some(event.bucket_id),
+                        )
+                        .await;
+
+                    let download_request = match download_request {
+                        Ok(download_request) => download_request,
+                        Err(error) => {
+                            error!(
+                                target: LOG_TARGET,
+                                "Failed to download chunk {:?} of file {:?} from peer {:?}: {:?}",
+                                chunk, file_key, peer_id, error
+                            );
+                            continue;
+                        }
+                    };
```

Comment:
Won't this fail quite often if we are requesting random BSPs for some data. Shouldn't we identify the BSPs that have this file first via the indexer and then request it from them?

Reply:
That's what we are doing. We query the indexer for all the files in a bucket, than for each file we query all the peer ids of BSPs storing that file and we cycle through them.